staging/src/k8s.io/client-go/rest/factory.go

```diff
@@ -0,0 +1,131 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rest
+
+import (
+	"net/http"
+	"net/url"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/util/flowcontrol"
+)
+
+type RESTClientFactory struct {
+	// base is the root URL for all invocations of the client
+	base *url.URL
+	// rateLimiter is shared among all requests created by this client unless specifically
+	// overridden.
+	rateLimiter flowcontrol.RateLimiter
+	// warningHandler is shared among all requests created by this client.
+	// If not set, defaultWarningHandler is used.
+	warningHandler WarningHandler
+	// creates BackoffManager that is passed to requests.
+	createBackoffMgr func() BackoffManager
+	// Set specific behavior of the client.  If not set http.DefaultClient will be used.
+	Client *http.Client
+}
+
+func RESTClientFactoryFor(config *Config) (*RESTClientFactory, error) {
+	// Base URL
+	base, _, err := defaultServerUrlFor(config)
+	if err != nil {
+		return nil, err
+	}
+
+	if !strings.HasSuffix(base.Path, "/") {
+		base.Path += "/"
+	}
+	base.RawQuery = ""
+	base.Fragment = ""
+
+	// Transport
+	if config.UserAgent == "" {
+		config.UserAgent = DefaultKubernetesUserAgent()
+	}
+
+	transport, err := TransportFor(config)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
the transport/client bits are really the only thing we need/want to reuse... the rest could be reconstructed

ideally, we wouldn't duplicate all the other stuff from the RESTClient constructor... it's vulnerable to drift

Reply:
transport + baseURL, right?
agree with the rest

Reply:
... the problem is that we miss options that are obtained from Config, if you construct the Factory from Config maybe is legit to expect to create the Clients with those

Reply:
baseURL isn't expensive to construct new per RESTClient, the transport is the only thing we care about sharing between all the RESTClients in the clientset, right?

I would try to structure the factory so that it *just* constructs the transport/client, then factor out RESTClient construction so that it does the same, and have a new rest client constructor that takes a config and a client

Reply:
maybe I'm overthinking this and we just need something like https://github.com/kubernetes/kubernetes/pull/105490/commits/225e32401e95e73532e2209c3a0eabef0a3482ed ?